### DISCUSSION TITLE
understaning how  where expression works

### DISCUSSION THREAD

**Discussion by iferencik** [developmentseed/titiler#544] [Q&A]
Hello folks,

I am facing  some difficulties understanding how the rendering works when using  where expressions.

I have a raster layer with values going from 1 to 6 (inclusive) [info](https://titiler.undpgeohub.org/cog/info?url=https%3A%2F%2Fundpngddlsgeohubdev01.blob.core.windows.net%2Fzero-hunger%2FCrop_Trends%2FCategorical_maize.tif%3Fc3Y9MjAyMS0xMC0wNCZzcz1iJnNydD1vJnNlPTIwMjMtMTEtMjRUMTklM0EzOSUzQTQxWiZzcD1yJnNpZz10Rk45cWdZOE5ZTVdUR3ozSHg5RGw4SGt0ZGE2aTVBMzdvU1duJTJCc1hMTjglM0Q%3D&bidx=1)
![image](https://user-images.githubusercontent.com/92793531/203863215-1cdfa3bc-0027-4925-af5e-c0382c02222b.png)

the nodata value is set to 0

for the image above i use following parameters to render the layer and it works well, the nodata(0) is transparent
```
scale: 1
TileMatrixSetId: WebMercatorQuad
url: https://undpngddlsgeohubdev01.blob.core.windows.net/zero-hunger/Crop_Trends/Categorical_maize.tif?c3Y9MjAyMS0xMC0wNCZzcz1iJnNydD1vJnNlPTIwMjMtMTEtMjRUMTklM0EzOSUzQTQxWiZzcD1yJnNpZz10Rk45cWdZOE5ZTVdUR3ozSHg5RGw4SGt0ZGE2aTVBMzdvU1duJTJCc1hMTjglM0Q=
bidx: 1
unscale: false
resampling: nearest
rescale: 1,6
return_mask: true
colormap_name: viridis

```

if i apply a where expression 

```
scale: 1
TileMatrixSetId: WebMercatorQuad
url: https://undpngddlsgeohubdev01.blob.core.windows.net/zero-hunger/Crop_Trends/Categorical_maize.tif?c3Y9MjAyMS0xMC0wNCZzcz1iJnNydD1vJnNlPTIwMjMtMTEtMjRUMTklM0EzOSUzQTQxWiZzcD1yJnNpZz10Rk45cWdZOE5ZTVdUR3ozSHg5RGw4SGt0ZGE2aTVBMzdvU1duJTJCc1hMTjglM0Q=
bidx: 1
unscale: false
resampling: nearest
rescale: 0,6
return_mask: true
colormap_name: viridis
expression: where(b1 > 3, b1, 0);
nodata: 0
```
![image](https://user-images.githubusercontent.com/92793531/203863796-d7a70196-3b11-4e57-8c74-217ef4f31e04.png)

The this is now the pixels having a value of 0 are no longer transparent, as I expected.
I am just curious if this behavior is correct.Is it possible to render the 0's transparent?

Thanks in advance

**Comment by vincentsarago** [comment]
@iferencik mask is not updated on data expression! https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/models.py#L430-L431

Hopefully this is the kind of use case for custom algorithm we are going to introduce in #542

**Reply by vincentsarago** [reply]
I think you could get around by using a discrete colormap, if you set `0: [0, 0, 0, 0]` it will create transparent pixel 

https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/utils.py#L419-L423

**Reply by iferencik** [reply]
@vincentsarago I see,

using discrete colormap works.

**Reply by iferencik** [reply]
@vincentsarago , 

if  [https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/models.py#L430](https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/models.py#L430]) line returns a numpy array, updating the mask would be easy 
i.e 
mask = np.where (apply_expression(blocks, self.band_names, self.data) == no_data, True, False).astype('u1')


would that work?

**Reply by vincentsarago** [reply]
No, because the dataset might not use Nodata and the ImageData class has no idea about it anyway! 

As I said, this will be quite trivial with the custom algorithm!

**Reply by iferencik** [reply]
do you have any docs/link relate to this algos? how are they supposed to work?

**Reply by iferencik** [reply]
guess the PR explains it :)

**Reply by vincentsarago** [reply]
> guess the PR explains it :)

well you have to dig a bit 

- https://github.com/developmentseed/titiler/blob/CustomAlgorithm/docs/src/advanced/Algorithms.md
- https://github.com/developmentseed/titiler/blob/CustomAlgorithm/docs/src/examples/code/tiler_with_custom_algorithm.md
- https://github.com/developmentseed/titiler/blob/CustomAlgorithm/docs/src/examples/notebooks/Working_with_Algorithm.ipynb

The PR is still WIP, but if you want to go through it, feel free to ask question/comment directly in #542

**Reply by iferencik** [reply]
thanks

**Comment by iferencik** [comment]
this actually looks fantastic :)

thanks @vincentsarago
